### PR TITLE
Python Module

### PR DESCRIPTION
**Pull request by airportyh**
## Why?

To support the Python template using Nix modules. This is meant to replicate the config from our @replit/Python template, which is spread out within `replit.nix`, `.replit`, `pyproject.toml` and `venv`. This new scheme will provide all of that from Nix out of the cacache disk.

## Changes

1. added a python module which includes nix build derivations for:
    a. custom pip
    b. custom poetry
    c. custom python-lsp-server
    d. replit-py package
    e. it also includes existing replitPackages: stderred, prybar-python310, debugpy, and dapPython
2. set and freeze LD_LIBRARY_PATH for the execution of python and prybar-python to include libraries needed for pandas, numpy, pygame and matplotlib via the creation of wrapper derivations
3. use `PIP_USER` env var to tell pip to install into a user local directory instead of inside the python installation. `PYTHONUSERBASE` tells it where the user local directory is
4. set `POETRY_VIRTUALENVS_CREATE` to instruct poetry not to create virtual environments
5. Read `pkgs/poetry/README.md` for more details as to how it works
6. lint fixes

## Testing

This is how I have been testing this:

1. `nix-build -A replitPackages.modules.python`
2. `cat result | jq .env`
3. set the variables in your local shell environment, adapt as needed. Example:
```
$ cat result|jq .env
{
  "PATH": "/nix/store/944vn6nqx6nibjj02v70k2fmvv628cbh-python3-wrapper/bin:/nix/store/8mi11ivaqwd9041jppy30n9h7ks5fbnh-python3.10-pip-21.2.dev0/bin:/nix/store/hwh8zyh0m3djl6hr9lmb676sy104d1ll-poetry/bin:/nix/store/mx0iaijf1sgk2bqm4d5xxlqq939yjpbj-prybar-python310-wrapper/bin:/nix/store/bav0j0149brw3hxxh4kp1xzx34s4yclq-python3.10-python-lsp-server-1.15.9/bin",
  "PIP_USER": "1",
  "POETRY_VIRTUALENVS_CREATE": "0",
  "PYTHONPATH": "$HOME/$REPL_SLUG/.pythonlibs/lib/python3.10/site-packages",
  "PYTHONUSERBASE": "$HOME/$REPL_SLUG/.pythonlibs"
}
$ export PATH="/nix/store/944vn6nqx6nibjj02v70k2fmvv628cbh-python3-wrapper/bin:/nix/store/8mi11ivaqwd9041jppy30n9h7ks5fbnh-python3.10-pip-21.2.dev0/bin:/nix/store/hwh8zyh0m3djl6hr9lmb676sy104d1ll-poetry/bin:/nix/store/mx0iaijf1sgk2bqm4d5xxlqq939yjpbj-prybar-python310-wrapper/bin:/nix/store/bav0j0149brw3hxxh4kp1xzx34s4yclq-python3.10-python-lsp-server-1.15.9/bin:$PATH"
$ export PIP_USER=1
$ export POETRY_VIRTUALENVS_CREATE=0
$ export PYTHONPATH="$HOME/.pythonlibs/lib/python3.10/site-packages"
$ export PYTHONUSERBASE="$HOME/.pythonlibs"
$ which poetry
/nix/store/hwh8zyh0m3djl6hr9lmb676sy104d1ll-poetry/bin/poetry
$ which pip
/nix/store/8mi11ivaqwd9041jppy30n9h7ks5fbnh-python3.10-pip-21.2.dev0/bin/pip
```
5. `mkdir test`
6. `cd test`
8. `poetry init` - go through answering the prompt and it will create pyproject.toml
9. `poetry add flask`
10. 
```
$ python3
> import flask
> flask
```
it should show the location of flask, which should be within the `.pythonlibs` directory
11. `poetry remove flask` should remove flask and its dependencies from the `.pythonlibs` directory
## Known Issues

These issues will need to be solved before merging:

* ~~currently using poetry to add an existing module (say flask) will result in strange behavior (it wants to update each module to the latest version, but can't, but yet does not error)~~ resolved
* ~~this change in pip https://github.com/replit/pip/tree/th-not-venv may or may not need to be merged~~
* ~~the `$HOME` and `$REPL_SLUG` env vars are not being rendered when used in a Nix modules, need to fix that~~ we'll fix this is pid1 later

### CODE REVIEW

**Comment by masad-frost** [comment]
Let's go!

File: modules/python.nix [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    # Needed for matplotlib
+    pkgs.xorg.libX11
+  ];
+
+  python3-wrapper = pkgs.stdenv.mkDerivation {
+    name = "python3-wrapper";
+    buildInputs = [ pkgs.makeWrapper ];
+    src = ./.;
+
+    installPhase = ''
```
Comment (ryantm) [inline modules/python.nix:38]:
```suggestion

    buildCommand = ''
```

File: modules/python.nix [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      ln -s $out/bin/python3 $out/bin/python3.10
+    '';
+  };
+
+  prybar-wrapper = pkgs.stdenv.mkDerivation {
+    name = "prybar-python310-wrapper";
+    buildInputs = [ pkgs.makeWrapper ];
+    src = ./.;
+
+    installPhase = ''
```
Comment (ryantm) [inline modules/python.nix:53]:
```suggestion

    buildCommand = ''
```

File: modules/swift.nix [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
+    src = ./.;
 
-  
-  installPhase = ''
-    mkdir -p $out/bin
-    makeWrapper ${pkgs.swift}/bin/swiftc $out/bin/swiftc \
-      --set PATH ""
-  '';
-};
+    installPhase = ''
```
Comment (ryantm) [inline modules/swift.nix:8]: get rid of src here too.

File: modules/python.nix [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    # Needed for pandas / numpy
+    pkgs.stdenv.cc.cc.lib
+    pkgs.zlib
+    # Needed for pygame
+    pkgs.glib
+    # Needed for matplotlib
+    pkgs.xorg.libX11
+  ];
+
+  python3-wrapper = pkgs.stdenv.mkDerivation {
```
Comment (ryantm) [inline modules/python.nix:33]:
```suggestion
  python3-wrapper = pkgs.stdenvNoCC.mkDerivation {
```
We don't need C compilers for this derivation.

File: modules/python.nix [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      mkdir -p $out/bin
+      makeWrapper ${python}/bin/python3 $out/bin/python3 \
+        --set LD_LIBRARY_PATH "${python-ld-library-path}"
+    
+      ln -s $out/bin/python3 $out/bin/python
+      ln -s $out/bin/python3 $out/bin/python3.10
+    '';
+  };
+
+  prybar-wrapper = pkgs.stdenv.mkDerivation {
```
Comment (ryantm) [inline modules/python.nix:48]:
```suggestion
  prybar-wrapper = pkgs.stdenvNoCC.mkDerivation {
```

File: modules/swift.nix [hunk 6]
```diff
@@ -1,18 +1,18 @@
 { pkgs, ... }:
-let swiftc-wrapper = pkgs.stdenv.mkDerivation {
-  name = "swiftc-wrapper";
-  buildInputs = [pkgs.makeWrapper];
-  src = ./.;
+let
+  swiftc-wrapper = pkgs.stdenv.mkDerivation {
```
Comment (ryantm) [inline modules/swift.nix:3]:
```suggestion
  swiftc-wrapper = pkgs.stdenvNoCC.mkDerivation {
```

File: pkgs/poetry/default.nix [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+      ${python}/bin/python3 -m venv $out/env
+      touch $out/env/poetry_env # This allows poetry to recognize it
+                                # https://github.com/replit/poetry/blob/replit-1.1/poetry/utils/env.py#L885
+                                # invoking the workaround so that poetry
+                                # does not use its own venv for the project
+                                # env
+
+      mkdir -p $out/poetry.bundle
+      tar xvf poetry.bundle.tar.gz -C $out/poetry.bundle
```
Comment (ryantm) [inline pkgs/poetry/default.nix:24]: Can we convert this to be a fixed-output-derivation (`pkgs.fetchTarball`), so we don't need to check in the bundle?

Comment (ryantm) [reply pkgs/poetry/default.nix:24]: Happy to pair with you to try to do this!

**Review by ryantm — CHANGES_REQUESTED** [review]
looks pretty good, some small suggestions and I'd be pretty sad to check in that 25MB tarball, let's see if we can do something about that.

File: modules/python.nix (hunk 5 above)
Comment (airportyh) [reply modules/python.nix:47]: :+1:

File: pkgs/poetry/default.nix (hunk 7 above)
Comment (airportyh) [reply pkgs/poetry/default.nix:24]: But then we'd have to publish the tarball somewhere?

File: modules/python.nix [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+  python-ld-library-path = pkgs.lib.makeLibraryPath [
+    # Needed for pandas / numpy
+    pkgs.stdenv.cc.cc.lib
+    pkgs.zlib
+    # Needed for pygame
+    pkgs.glib
+    # Needed for matplotlib
+    pkgs.xorg.libX11
+  ];
```
Comment (lhchavez) [inline modules/python.nix:29]: i long for the day when this is no longer (explicitly) needed

File: modules/python.nix [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+  python3-wrapper = pkgs.stdenvNoCC.mkDerivation {
+    name = "python3-wrapper";
+    buildInputs = [ pkgs.makeWrapper ];
+
+    buildCommand = ''
+      mkdir -p $out/bin
+      makeWrapper ${python}/bin/python3 $out/bin/python3 \
+        --set LD_LIBRARY_PATH "${python-ld-library-path}"
```
Comment (lhchavez) [inline modules/python.nix:38]: what does this do? does it set the `LD_LIBRARY_PATH` env variable every time one runs `python3`?

some things broke when we did that outside of nix, but given that this is being done _inside_ it might be different. how are we testing this?

File: modules/python.nix [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    
+      ln -s $out/bin/python3 $out/bin/python
+      ln -s $out/bin/python3 $out/bin/python3.10
```
Comment (lhchavez) [inline modules/python.nix:41]: oof, this might bite us in the future: minor versions of python are not mutually-compatible, since sometimes Python maintainers _remove_ old functionality: we may need to call this module `python_3_10`.

File: modules/python.nix (hunk 8 above)
Comment (airportyh) [reply modules/python.nix:29]: same

File: modules/python.nix (hunk 9 above)
Comment (airportyh) [reply modules/python.nix:38]: It creates this shell script:
```
#! /nix/store/dsd5gz46hdbdk2rfdimqddhq6m8m8fqs-bash-5.1-p16/bin/bash -e
export LD_LIBRARY_PATH='/nix/store/mdck89nsfisflwjv6xv8ydj7dj0sj2pn-gcc-11.3.0-lib/lib:/nix/store/026hln0aq1hyshaxsdvhg0kmcm6yf45r-zlib-1.2.13/lib:/nix/store/2k366jrbsra97gjfxwvrhvixjfxdach5-glib-2.74.1/lib:/nix/store/w3zzhfl4a7xp0xfflz2gawv02y8ba9z8-libX11-1.8.1/lib'
exec "/nix/store/hd4cc9rh83j291r5539hkf6qd8lgiikb-python3-3.10.8/bin/python3"  "$@"
```

Comment (airportyh) [reply modules/python.nix:38]: Yeah, I know the bug. This is isolated from the user's environment, so it should be fine.

File: pkgs/poetry/poetry-bundle.nix [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    owner = "replit";
+    repo = "poetry";
+    rev = "replit-1.1";
+    sha256 = "sha256-KKm6+i8ilPRx19wNaeDjrypdLCqgYE1JKwt2USx7HHk=";
+  };
+
+  buildInputs = [ pkgs.python310Packages.pip pkgs.toml2json pkgs.jq ];
+
+  buildPhase = ''
+    ls -la
```
Comment (ryantm) [inline pkgs/poetry/poetry-bundle.nix:14]:
```suggestion
```

**Review by ryantm — CHANGES_REQUESTED** [review]
some nits

File: pkgs/poetry/scripts/fetch-bundle.sh [hunk 12]
```diff
@@ -0,0 +1,21 @@
+#! /usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (airportyh) [reply pkgs/poetry/scripts/fetch-bundle.sh:1]: :+1:

**Review by ryantm — APPROVED** [review]
looks good to me now

File: modules/python.nix (hunk 9 above)
Comment (lhchavez) [reply modules/python.nix:38]: i suspect that it still may cause issues: if there's a DSO bundled into a package that is linked against another version of glibc we may see linker errors. if we test enough edge cases where there are `.so` files in the wheel, we're good.

**Comment by lhchavez** [comment]
one last q: when users install packages, where are they installed? are we secretly creating a `venv` directory in the Repl somehow?

**Review by masad-frost — APPROVED** [review]
Nothing from my end :shipit: and we can iterate

**Review by ankit-gautam23 — COMMENTED** [review]
LGTM
